File: 2015/Day02/Part1/python/task1.py
# HU: A helyes végeredmény, dokumentációs céllal.
# EN: The correct final answer, for documentation purposes.
# 1586300 - That's the right answer!

# HU: A teljes szükséges papírmennyiséget tároló változó inicializálása.
# EN: Initializing the variable that will store the total required amount of paper.
total = 0

# HU: Ideiglenes változók az oldalak területeinek. Ezeket a ciklus minden körben felülírja.
# EN: Temporary variables for the side areas. These are overwritten in each loop iteration.
part1 = 0
part2 = 0
part3 = 0

l = 0
w = 0
h = 0

# HU: Fájl megnyitása olvasásra a biztonságos 'with' blokkban, ami automatikusan bezárja a fájlt.
# EN: Opening the file for reading in a safe 'with' block, which closes the file automatically.
with open('input.txt', 'r') as file:
    # HU: Soronként végigmegyünk a fájlon. Minden sor egy ajándék méreteit tartalmazza.
    # EN: We loop through the file line by line. Each line contains the dimensions of one present.
    for line in file:
        # HU: Eltávolítjuk a felesleges szóközöket és az 'x' mentén feldaraboljuk a sort egy listába.
        # EN: We remove unnecessary whitespace and split the line by the 'x' character into a list.
        box = line.strip().split('x')
        
        # HU: A szövegként beolvasott méreteket egésszé (integer) alakítjuk.
        # EN: We convert the dimensions, read as text, into integers.
        l = int(box[0])
        w = int(box[1])
        h = int(box[2])

        # HU: Kiszámoljuk a doboz három különböző oldalának területét.
        # EN: We calculate the area of the three unique sides of the box.
        part1 = l*w
        part2 = w*h
        part3 = h*l

        # HU: Kiszámoljuk a doboz teljes felszínét a képlet (2*lw + 2*wh + 2*hl) alapján.
        # EN: We calculate the total surface area of the box based on the formula (2*lw + 2*wh + 2*hl).
        surface_area = 2*(part1+part2+part3)

        # HU: A teljes összeghez hozzáadjuk a felszínt és a ráadást (a legkisebb oldal területe).
        # EN: We add the surface area and the slack (the area of the smallest side) to the running total.
        total += surface_area + min(part1, part2, part3)

# HU: Kiírjuk a végső, összesített papírmennyiséget.
# EN: We print the final, total amount of paper.
print(total)

File: 2015/Day02/Part1/cpp/task1.cpp
// HU: Egy nem-sztenderd, kényelmi header, ami szinte minden C++ standard könyvtárat tartalmaz.
// EN: A non-standard, convenience header that includes almost every C++ standard library.
#include <bits/stdc++.h>

// HU: Lehetővé teszi, hogy ne kelljen a std:: előtagot használni (pl. cout, string).
// EN: Allows using standard library components without the std:: prefix (e.g., cout, string).
using namespace std;

int main() {
    // HU: Változók inicializálása. Fontos minden változónak kezdőértéket adni.
    // EN: Initializing variables. It's important to give an initial value to every variable.
    int total = 0;
    int l=0, w=0, h=0;
    int side1=0, side2=0, side3=0;
    int surface_area = 0;

    // HU: Megnyitjuk az input.txt fájlt olvasásra.
    // EN: We open the input.txt file for reading.
    ifstream file("input.txt");

    // HU: Egy string változó a fájl egy sorának tárolására.
    // EN: A string variable to store one line of the file.
    string line;

    // HU: Ellenőrizzük, hogy a fájl megnyitása sikeres volt-e.
    // EN: We check if the file was opened successfully.
    if (file.is_open()) {
        // HU: Soronként végigmegyünk a fájlon, amíg van olvasható sor.
        // EN: We loop through the file line by line, as long as there are lines to read.
        while (getline(file, line)) {
            // HU: Létrehozunk egy stringstream objektumot az aktuális sorból a könnyebb darabolásért.
            // EN: We create a stringstream object from the current line for easier parsing.
            stringstream ss(line);

            // HU: Ideiglenes string változók a darabolt részeknek.
            // EN: Temporary string variables for the parsed parts.
            string l_str, w_str, h_str;
            
            // HU: A getline-t használjuk a 'ss' streamen, 'x' határoló karakterrel.
            // EN: We use getline on the 'ss' stream with 'x' as the delimiter.
            getline(ss, l_str, 'x');
            getline(ss, w_str, 'x');
            getline(ss, h_str); // HU: Az utolsó darab a sor végéig tart. / EN: The last part goes until the end of the line.

            // HU: A beolvasott szövegeket (string) számokká (integer) alakítjuk a 'stoi' függvénnyel.
            // EN: We convert the read texts (strings) into numbers (integers) with the 'stoi' function.
            l = stoi(l_str);
            w = stoi(w_str);
            h = stoi(h_str);

            // HU: Kiszámoljuk a doboz három különböző oldalának területét.
            // EN: We calculate the area of the three unique sides of the box.
            side1 = l*w;
            side2 = w*h;
            side3 = h*l;

            // HU: Kiszámoljuk a doboz teljes felszínét.
            // EN: We calculate the total surface area of the box.
            surface_area = 2 * (side1 + side2 + side3);

            // HU: A teljes összeghez hozzáadjuk a felszínt és a ráadást (a legkisebb oldal területe).
            // EN: We add the surface area and the slack (the area of the smallest side) to the running total.
            // HU: A min() függvényhez kapcsos zárójeleket használunk, hogy több elemből válasszon.
            // EN: We use curly braces for the min() function to choose from multiple elements.
            total += surface_area + min({side1, side2, side3});
        }

        // HU: Bezárjuk a fájlt.
        // EN: We close the file.
        file.close();

        // HU: Kiírjuk a végső, összesített papírmennyiséget.
        // EN: We print the final, total amount of paper.
        cout << total << endl;
    }
    else {
        // HU: Hibaüzenet, ha a fájl nem nyitható meg.
        // EN: Error message if the file cannot be opened.
        cerr << "Unable to open file" << endl;
    }

    // HU: A program sikeresen lefutott.
    // EN: The program executed successfully.
    return 0;
}